contracts/contracts/util.cairo

```diff
@@ -0,0 +1,44 @@
+from starkware.cairo.common.math import assert_not_zero
+from starkware.cairo.common.default_dict import default_dict_new, default_dict_finalize
+from starkware.cairo.common.dict import dict_update
+from starkware.cairo.common.dict_access import DictAccess
+
+func almost_equal(a, b) -> (res : felt):
+    if (a - b) * (a - b - 1) * (a - b + 1) == 0:
+        return (res=1)
+    end
+    return (res=0)
+end
+
+func assert_unique_elements_impl{range_check_ptr, dict_ptr : DictAccess*}(
+    index : felt, data_len : felt, data : felt*
+):
+    alloc_locals
+    if index == data_len:
+        return ()
+    end
+
+    # Fails if some key's value was set
+    dict_update(key=[data], prev_value=data_len, new_value=index)
+
+    return assert_unique_elements_impl(index + 1, data_len, data + 1)
+end
+
+func assert_unique_elements{range_check_ptr}(data_len : felt, data : felt*):
+    alloc_locals
+
+    # default_value shall be not less than data_len
+    # otherwise data[default_value] can have duplicates for index in (default_value, data_len)
+    let (local my_dict_start) = default_dict_new(default_value=data_len)
+    let my_dict_end = my_dict_start
+
+    assert_unique_elements_impl{range_check_ptr=range_check_ptr, dict_ptr=my_dict_end}(
+        index=0, data_len=data_len, data=data
+    )
+
+    let (finalized_dict_start, finalized_dict_end) = default_dict_finalize(
```

Comment:
Hmm, I wonder if this is needed. The values are not used anyway - this function only needs to revert if there are duplicates, but otherwise I don't think it needs to be setting variables

Reply:
I pointed this out too but I wasn't sure if it's safe:

> Note that in order to enforce the consistency of subsequent dictionary accesses with the default values, one must eventually call default_dict_finalize() (which in turn calls dict_squash(), as discussed in the corresponding section). Otherwise, this is only enforced by hints which can be bypassed by a malicious prover.

Reply:
Ah, interesting. Please at least add a link to the explanation next to the code.

Reply:
Removed unused return values. Added link and comment